### PR TITLE
JAMES-2555 Webadmin APIs for message reIndexing

### PR DESCRIPTION
**Pull request by chibenwa**
Note that message reIndexing per messageId will be added in another PR

### CODE REVIEW

**Comment by rouazana** [comment]
```
[d8bc79483142126dc8a3f935576b3be6beba12fe] [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.091 s <<< FAILURE! - in org.apache.james.JamesCapabilitiesServerTest
[d8bc79483142126dc8a3f935576b3be6beba12fe] [ERROR] startShouldSucceedWhenRequiredCapabilities{GuiceJamesServer}  Time elapsed: 2.067 s  <<< ERROR!
[d8bc79483142126dc8a3f935576b3be6beba12fe] com.google.inject.CreationException: 
[d8bc79483142126dc8a3f935576b3be6beba12fe] Unable to create injector, see the following errors:
[d8bc79483142126dc8a3f935576b3be6beba12fe] 
[d8bc79483142126dc8a3f935576b3be6beba12fe] 1) No implementation for org.apache.james.mailbox.store.search.ListeningMessageSearchIndex was bound.
[d8bc79483142126dc8a3f935576b3be6beba12fe]   while locating org.apache.james.mailbox.store.search.ListeningMessageSearchIndex
[d8bc79483142126dc8a3f935576b3be6beba12fe]     for the 2nd parameter of org.apache.mailbox.tools.indexer.ReIndexerPerformer.<init>(ReIndexerPerformer.java:65)
[d8bc79483142126dc8a3f935576b3be6beba12fe]   while locating org.apache.mailbox.tools.indexer.ReIndexerPerformer
[d8bc79483142126dc8a3f935576b3be6beba12fe]     for the 1st parameter of org.apache.mailbox.tools.indexer.ReIndexerImpl.<init>(ReIndexerImpl.java:44)
[d8bc79483142126dc8a3f935576b3be6beba12fe]   at org.apache.james.modules.mailbox.JPAMailboxModule.configure(JPAMailboxModule.java:86) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.mailbox.JPAMailboxModule)
```

**Comment by chibenwa** [comment]
Was :green_apple: 

I will force push to solve conflict

File: src/site/markdown/server/manage-webadmin.md [hunk 1]
```diff
@@ -2041,6 +2042,198 @@ curl -XDELETE http://ip:port/sieve/quota/users/user@domain.com
...
+ - Location header indicates the location of the resource associated with the scheduled task. Example:
+
+```
+Location: /tasks/3294a976-ce63-491e-bd52-1b6f465ed7a2
+```
+
+Response codes:
+
+ - 201: Success. Corresponding task id is returned.
+ - 400: Error in the request. Details can be found in the reported error.
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: I saw in Swagger doc, there is one more response code: 500 documented.

File: src/site/markdown/server/manage-webadmin.md [hunk 2]
```diff
@@ -2041,6 +2042,198 @@ curl -XDELETE http://ip:port/sieve/quota/users/user@domain.com
...
+
+The scheduled task will have the following type `FullReIndexing` and the following `additionalInformation`:
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: s/FullReIndexing/UserReIndexing (not sure that exactly)

File: src/site/markdown/server/manage-webadmin.md [hunk 3]
```diff
@@ -2041,6 +2042,198 @@ curl -XDELETE http://ip:port/sieve/quota/users/user@domain.com
...
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: a/FullReIndexing/MailboxReIndexing (not sure that exactly)

File: src/site/markdown/server/manage-webadmin.md [hunk 4]
```diff
@@ -2041,6 +2042,198 @@ curl -XDELETE http://ip:port/sieve/quota/users/user@domain.com
...
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: s/FullReIndexing/MessageReIndexing (not sure that exactly)

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 5]
```diff
@@ -60,7 +60,7 @@
     private final MailboxSessionMapperFactory mailboxSessionMapperFactory;
 
     @Inject
-    ReIndexerPerformer(MailboxManager mailboxManager,
+    public ReIndexerPerformer(MailboxManager mailboxManager,
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java]: code formating for below linebreak

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+            }
+        }
+    }
+
+    @Nested
+    class UserReprocessing {
+        @Nested
+        class Validation {
+            @Test
+            void userlReprocessingShouldFailWithNoTask() {
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java]: s/userl/user

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java [hunk 7]
```diff
@@ -0,0 +1,708 @@
...
+                    .post("/index/user/" + USERNAME + "/mailbox/bad?task=reIndex")
+                .then()
+                    .statusCode(HttpStatus.BAD_REQUEST_400)
+                    .body("statusCode", is(400))
+                    .body("type", is(ErrorResponder.ErrorType.INVALID_ARGUMENT.getType()))
+                    .body("message", is("Error while parsing 'mailbox'"));
+            }
+
+            @Test
+            void mailboxReprocessingShouldFailWithNonExistantMailboxId() {
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java]: s/Existant/Existent

Comment (trantienduchn) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java]: and how about one more `mailboxReprocessingShouldFailWithNonExistentUser`

File: src/site/markdown/server/manage-webadmin.md (hunk 1 above)
Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: It is implicit in webadmin website doc: see the top of the file....

**Comment by chibenwa** [comment]
```
[510991158d29214d8f5af9f518dfb25612918d84] [ERROR]   RabbitMQClusterTest$ClusterSharing.queuesShouldBeShared » Completion org.testc...
[510991158d29214d8f5af9f518dfb25612918d84] [ERROR]   RabbitMQClusterTest$ClusterSharing.rabbitMQManagerShouldReturnThreeNodesWhenAskingForStatus » Completion
```

test this please

**Comment by chibenwa** [comment]
```
[510991158d29214d8f5af9f518dfb25612918d84] Location: /org/apache/james/imap/scripts/ValidateDeployment.test:499
[510991158d29214d8f5af9f518dfb25612918d84] LastClientMsg: c SEARCH FLAGGED
[510991158d29214d8f5af9f518dfb25612918d84] Expected: '\* SEARCH 1 2 3 4'
[510991158d29214d8f5af9f518dfb25612918d84] Actual   : '* SEARCH 1 2 3'
```

In Cassandra deployment test

**Comment by chibenwa** [comment]
test this please

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+import io.swagger.annotations.ApiImplicitParams;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import spark.Request;
+import spark.Response;
+import spark.Service;
+
+@Api(tags = "ReIndexing")
+@Path("/mailboxIndex")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java]: how mailboxIndex is different than mailbox ?
I mean, is mailboxindex a resource of its own ?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side."),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - details in the returned error message")
+    })
+    private TaskIdDto reIndexAll(Request request, Response response) {
+        return wrap(request, response,
+            () -> reIndexer.reIndex());
+    }
+
+    @POST
+    @Path("/user/{user}")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java]: s/user/users : when you expect that something contains a resource (a given user) the path at the previous level is supposed to be a plural (and can allow to list resources)

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side."),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - details in the returned error message")
+    })
+    private TaskIdDto reIndexMailbox(Request request, Response response) {
+        return wrap(request, response,
+            () -> reIndexer.reIndex(retrievePath(request)));
+    }
+
+    @POST
+    @Path("/user/{user}/mailbox/{mailboxId}/uid/{uid}")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java]: `uid` is not a resource, maybe you mean `mail`

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java (hunk 8 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRoutes.java]: I more see it as a path prefix to subsequent resources.

**Comment by chibenwa** [comment]
```
[c68a6be840abb44571a25d254b1a5c6183c48e37] [ERROR]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxWhenSameMessageHasAlreadyBeenMovedToSpam:136 » ConditionTimeout
```

test this please

**Comment by chibenwa** [comment]
Rebased to ease review.

Was :green_apple:

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java [hunk 11]
```diff
@@ -0,0 +1,719 @@
...
+                    .post("/mailboxIndex/users/" + USERNAME + "/mailboxes/bad/mails/7?task=reIndex")
+                .then()
+                    .statusCode(HttpStatus.BAD_REQUEST_400)
+                    .body("statusCode", is(400))
+                    .body("type", is(ErrorResponder.ErrorType.INVALID_ARGUMENT.getType()))
+                    .body("message", is("Error while parsing 'mailbox'"));
+            }
+
+            @Test
+            void messageReprocessingShouldFailWithNonExistantMailboxId() {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java]: s/existant/existent everywhere in this file

**Comment by chibenwa** [comment]
Was :green_apple:

**Comment by chibenwa** [comment]
Was :green_apple:

**Comment by chibenwa** [comment]
Was :green_apple:

**Comment by chibenwa** [comment]
Was :green_apple:

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**